src/external/wannier_shared.jl

```diff
@@ -0,0 +1,407 @@
+                    @printf fp "%25.18f %25.18f\n" real(ψnk_real[ix, iy, iz]) imag(ψnk_real[ix, iy, iz])
+                end
+            end
+        end
+    end
+    nothing
+end
+
+
+@doc raw"""
+Computes the matrix ``[M^{k,b}]_{m,n} = \langle u_{m,k} | u_{n,k+b} \rangle``
+for given `k`, `kpb` = ``k+b``.
+
+`G_shift` is the "shifting" vector, correction due to the periodicity conditions
+imposed on ``k \to  ψ_k``.
+It is non zero if `kpb` is taken in another unit cell of the reciprocal lattice.
+We use here that:
+``u_{n(k + G_{\rm shift})}(r) = e^{-i*\langle G_{\rm shift},r \rangle} u_{nk}``.
+"""
+@views function overlap_Mmn_k_kpb(basis::PlaneWaveBasis, ψ, ik, ikpb, G_shift, n_bands)
+    # Search for common Fourier modes and their resp. indices in Bloch states k and kpb
+    # TODO Check if this can be improved using the G vector mapping in the kpoints
+    k   = basis.kpoints[ik]
+    kpb = basis.kpoints[ikpb]
+    equivalent_G_vectors = [(iGk, index_G_vectors(basis, kpb, Gk + G_shift))
+                            for (iGk, Gk) in enumerate(G_vectors(basis, k))]
+    iGk   = [eqG[1] for eqG in equivalent_G_vectors if !isnothing(eqG[2])]
+    iGkpb = [eqG[2] for eqG in equivalent_G_vectors if !isnothing(eqG[2])]
+
+    # Compute overlaps
+    # TODO This should be improved ...
+    Mkb = zeros(ComplexF64, (n_bands, n_bands))
+    for n in 1:n_bands
+        for m in 1:n_bands
+            # Select the coefficient in right order
+            Mkb[m, n] = dot(ψ[ik][iGk, m], ψ[ikpb][iGkpb, n])
+        end
+    end
+    iszero(Mkb) && return Matrix(I, n_bands, n_bands)
+    Mkb
+end
+
+@timing function write_w90_mmn(fileprefix::String, basis::PlaneWaveBasis, ψ, nnkp; n_bands)
+    open(fileprefix * ".mmn", "w") do fp
+        println(fp, "Generated by DFTK at $(now())")
+        println(fp, "$n_bands  $(length(ψ))  $(nnkp.nntot)")
+        for (ik, ikpb, G_shift) in nnkp.nnkpts
+            @printf fp "%i  %i  %i  %i  %i \n" ik ikpb G_shift...
+            for ovlp in overlap_Mmn_k_kpb(basis, ψ, ik, ikpb, G_shift, n_bands)
+                @printf fp "%22.18f %22.18f \n" real(ovlp) imag(ovlp)
+            end
+        end
+    end
+    nothing
+end
+
+@doc raw"""
+Compute the matrix ``[A_k]_{m,n} = \langle ψ_m^k | g^{\text{per}}_n \rangle``
+
+``g^{per}_n`` are periodized gaussians whose respective centers are given as an
+ (num_bands,1) array [ [center 1], ... ].
+
+Centers are to be given in lattice coordinates and G_vectors in reduced coordinates.
+The dot product is computed in the Fourier space.
+
+Given an orbital ``g_n``, the periodized orbital is defined by :
+ ``g^{per}_n =  \sum\limits_{R \in {\rm lattice}} g_n( \cdot - R)``.
+The  Fourier coefficient of ``g^{per}_n`` at any G
+is given by the value of the Fourier transform of ``g_n`` in G.
+"""
+function compute_amn_kpoint(basis::PlaneWaveBasis, ψk, kpt, projections::AbstractVector{<:WannierProjection}, n_bands)
```

Comment:
in internal function names we typically abbreviate `kpoint` as `kpt`
Typically we have the `kpt` argument follow the `basis`
Don't annotate the type of the projections (only `AbstractVector`)
Make `n_bands` a kwarg like in the other functions.

Side note: If we do a restriction of the ψk arrays inside the ψ vector itself using a view we could rid of passing both the ψk and the n_bands. Here is a code example that does this using masks:
https://github.com/JuliaMolSim/DFTK.jl/blob/master/src/response/chi0.jl#L348

Reply:
I just moved `kpt` to the second position for now. (We could move to an array of views but it makes the code less readable IMO and would need to be applied to the other functions as well.)

Reply:
ok